src/sqlfluff/core/parser/parsers.py

```diff
@@ -0,0 +1,179 @@
+        type: Optional[str] = None,
+        optional: bool = False,
+        **segment_kwargs,
+    ):
+        # String matchers are not case sensitive, so we make the template
+        # uppercase on creation. If any SQL dialect is found to be case
+        # sensitive for keywords, this could be extended to allow
+        # case sentivitiy.
+        self.template = template.upper()
+        self.raw_class = raw_class
+        self.name = name
+        self.type = type
+        self.optional = optional
+        self.segment_kwargs = segment_kwargs or {}
+
+    def is_optional(self) -> bool:
+        """Return whether this element is optional."""
+        return self.optional
+
+    def simple(self, parse_context: "ParseContext") -> Optional[List[str]]:
+        """Return simple options for this matcher.
+
+        Because string matchers are not case sensitive we can
+        just return the template here.
+        """
+        return [self.template]
+
+    def _is_first_match(self, segment: BaseSegment):
+        """Does the segment provided match according to the current rules."""
+        # Is the target a match and IS IT CODE.
+        # The latter stops us accidentally matching comments.
+        if self.template == segment.raw.upper() and segment.is_code:
+            return True
+        return False
+
+    def _make_match_from_first_result(self, segments: Tuple[BaseSegment, ...]):
+        """Make a MatchResult from the first segment in the given list.
+
+        This is a helper function for reuse by other parsers.
+        """
+        # Construct the segment object
+        new_seg = self.raw_class(
+            raw=segments[0].raw,
+            pos_marker=segments[0].pos_marker,
+            type=self.type,
+            name=self.name,
+            **self.segment_kwargs,
+        )
+        # Return as a tuple
+        return MatchResult((new_seg,), segments[1:])
+
+    def match(
+        self,
+        segments: Union[BaseSegment, Tuple[BaseSegment, ...]],
+        parse_context: "ParseContext",
+    ) -> MatchResult:
+        """Compare input segments for a match, return a `MatchResult`.
+
+        Note: For matching here, we only consider the *first* element,
+        because we assume that a keyword can only span one raw segment.
+        """
+        # If we've been passed the singular, make it a tuple
+        if isinstance(segments, BaseSegment):
+            segments = (segments,)
+
+        # We're only going to match against the first element
+        if len(segments) >= 1:
+            if self._is_first_match(segments[0]):
+                return self._make_match_from_first_result(segments)
+        return MatchResult.from_unmatched(segments)
+
+
+class NamedParser(StringParser):
+    """An object which matches and returns raw segments based on names."""
+
+    def simple(cls, parse_context: ParseContext) -> Optional[List[str]]:
+        """Does this matcher support a uppercase hash matching route?

```

Comment:
It's the shortcut used during matching so that the parser can go faster. It uses uppercase versions of segments and uses native string methods to look for locations ahead in the file.